cert/cert.go

```diff
@@ -188,3 +180,163 @@ func bigIntHash(n *big.Int) []byte {
 	h.Write(n.Bytes())
 	return h.Sum(nil)
 }
+
+// KeyType for RSA/ECDSA
+type KeyType uint8
+
+// OtherName type for asn1 encoding
+type OtherName struct {
+	A string `asn1:"utf8"`
+}
+
+// GeneralName type for asn1 encoding
+type GeneralName struct {
+	OID       asn1.ObjectIdentifier
+	OtherName `asn1:"tag:0"`
+}
+
+// GeneralNames type for asn1 encoding
+type GeneralNames struct {
+	GeneralName `asn1:"tag:0"`
+}
+
+var (
+	// https://support.microsoft.com/en-us/kb/287547
+	//  szOID_NT_PRINCIPAL_NAME 1.3.6.1.4.1.311.20.2.3
+	szOID = asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 311, 20, 2, 3}
+	// http://www.umich.edu/~x509/ssleay/asn1-oids.html
+	// 2 5 29 17  subjectAltName
+	subjAltName = asn1.ObjectIdentifier{2, 5, 29, 17}
+)
+
+//  getUPNExtensionValue returns marsheled asn1 encoded info
+func getUPNExtensionValue(subject pkix.Name) ([]byte, error) {
+	// returns the ASN.1 encoding of val
+	// in addition to the struct tags recognized
+	// we used:
+	// utf8 => causes string to be marsheled as ASN.1, UTF8 strings
+	// tag:x => specifies the ASN.1 tag number; imples ASN.1 CONTEXT SPECIFIC
+	return asn1.Marshal(GeneralNames{
+		GeneralName: GeneralName{
+			// init our ASN.1 object identifier
+			OID: szOID,
+			OtherName: OtherName{
+				A: subject.CommonName,
+			},
+		},
+	})
+}
+
+// NewClientCert generates a new x509 certificates based on the CertConfig passed
+func NewClientCert(commonName, UUID string, expiry time.Time) ([]byte, []byte, error) {
```

Comment:
AFAICS, the only difference we care about between NewClientCert and NewLeaf is that NewClientCert adds the UPN extension. So to avoid repeating all the generic certificate generation code here, I'd rather we used the NewLeaf to generate the cert, and provide the UPN extension as an argument to NewLeaf.